text/0000-attributes-multi.md

```diff
@@ -0,0 +1,229 @@
+expanding to those any number of attributes when the predicate is true. This
+was the first place to allow multiple attributes together with commas. This
+RFC is the second place. Outside of arbitrary macros, this RFC suggests doing
+so in the only other place we allow multiple attributes together.
+
+## Attribute Syntax
+
+In [RFC 2539], we defined the attribute syntax and a restriction on attributes
+so that commas cannot end up in the top level of attribute syntax. This means
+that attribute syntax is as follows:
+
+An attribute can look like:
+
+* `name`,
+* ``name(`TokenStream`)``
+* ``name = `TokenTree` ``
+
+where `TokenStream` is a sequence of tokens that only has the restriction that
+delimiters match and `TokenTree` is a single identifer, literal, punctuation
+mark, or a delimited `TokenStream`.
+
+The following cannot ever be parsed as attributes:
+
+* `name, option`
+* `name = some, options`
+
+Arguably, we could allow `(name, option)`, but we shouldn't.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Attribute containers may contain multiple attributes, delimited by commas. They
+are resolved as if each attribute was in its own attribute brace. Furthermore,
+attribute containers allow a trailing comma and may contain zero attributes.
+
+There is no behavior difference between putting each attribute in its own
+container and putting the attributes in the same container. For example, given
+three attributes, `attr1`, `attr2`, and `attr3`, the following will all behave
+identically:
+
+* `#[attr1] #[attr2] #[attr3]`
+* `#[attr1, attr2] #[attr3]`
+* `#[attr1] #[attr2, attr3]`
+* `#[attr1, attr2, attr3]`
+
+For a real example of multiple attributes in the same container, we have a
+test that is ignored:
+
+```rust
+#[test, ignore]
+fn should_not_panic() {
+    panic!("Oh nooooooooo");
+}
+```
+
+## Attribute Macro Input `TokenStream` Changes
+ 
+Before this RFC, when an attriute macro is executed, it is passed two
+`TokenStream`s. The first is the input passed to the attribute itself. That is
+unchanged by this RFC. The second is the `TokenStream` of the syntactic element
+the attribute is attached to without the attribute macro. Because the attribute
+macro was the only attribute in the attribute container, this `TokenStream`
+did not have the attribute container at all. This behavior continues as is,
+but we add that in the case of the attribute container containing multiple
+attributes, only the attribute macro and its following comma, if it exists, is
+removed. This is easy to see by example:
+
+Source | Macro Attribute `TokenStream` as String
+--- | ---
+`#[macro_attr] thing` | `"thing"`
+`#[attr1] #[macro_attr] thing` | `"#[attr1] thing"`
+`#[attr1, macro_attr] thing` | `"#[attr1, ] thing"`
+`#[macro_attr, attr1] thing` | `"#[attr1] thing"`
+`#[attr1, macro_attr, attr2] thing` | `"#[attr1, attr2] thing"`
+`#[] #[macro_attr] thing` | `"#[] thing"` or `"thing"` (see unresolved questions)
```

Comment:
I'm wary of guaranteeing contents of input streams with such precision (at least right now).

For example, we don't specify what `cfg_attr(predicate, a, b)` unwraps into - `#[a, b]` or `#[a] #[b]`, but now it makes observable difference (it would be reasonable to expect the former, but it's actually the latter).
Also, it may be convenient to split all multi-attribute containers into single-attribute containers in the initial implementation.

Macro authors should not rely on differences between `#[a] #[b]` and `#[a, b]` in general, and should support both, so we don't get much from specifying precise tokens.

Reply:
> Macro authors should not rely on differences between #[a] #[b] and #[a, b]

When a macro author highlights the span of the attribute item containing e.g. `#[a, b,]` from `#[a, b, c]` where some `c` has been removed, what span does that point to? `a, b` without the `#[]` or the `c`? Similarly, does `#[a]` from `#[a, b, c]` refer to just the `a`? I'd really like to see the exact semantics of whatever attribute modification and span transformation we do specified, both for the sake of this RFC and for https://github.com/rust-lang/rfcs/pull/2539.